lib/containers/engine.pm

```diff
@@ -0,0 +1,396 @@
+}
+
+=head2 remove_container
+
+Remove a container from the pool.
+
+=cut
+sub remove_container {
+    my ($self, $container_name) = @_;
+    $self->_engine_assert_script_run("rm -f $container_name");
+}
+
+=head2 check_image_in_host
+
+Returns true if host contains C<img> or false.
+
+=cut
+sub check_image_in_host {
+    my ($self, $img) = @_;
+    my @lregistry = $self->enum_images();
+    $img =~ @lregistry;
+}
+
+=head2 configure_insecure_registries
+
+Updates the registry files for the running container runtime to allow access to
+insecure registries.
+
+Implementation is subject to the the subclass otherwise call to this subroutine dies.
+
+=cut
+sub configure_insecure_registries {
+    my ($self) = shift;
+}
+
+=head2 cleanup_system_host
+
+Remove containers and then all the images respectively.
+Asserts that everything was cleaned up unless c<assert> is set to 0.
+
+=cut
+sub cleanup_system_host {
+    my ($self, $assert) = @_;
+    $assert // 1;
+    $self->_engine_assert_script_run("ps -q | xargs -r " . $self->runtime . " stop", 180);
+    $self->_engine_assert_script_run("system prune -a -f",                           180);
+
+    if ($assert) {
+        assert_equals(0, scalar @{$self->enum_containers()}, "containers have not been removed");
+        assert_equals(0, scalar @{$self->enum_images()},     "images have not been removed");
+    }
+}
+
+# Container engine subclasses here
+package containers::engine::docker;
+use Mojo::Base 'containers::engine';
+use testapi;
+use containers::utils qw(registry_url);
+use utils qw(systemctl file_content_replace);
+has runtime => 'docker';
+
+sub configure_insecure_registries {
+    my ($self) = shift;
+    my $registry = registry_url();
+    # Allow our internal 'insecure' registry
+    assert_script_run(
+        'echo "{ \"debug\": true, \"insecure-registries\" : [\"localhost:5000\", \"registry.suse.de\", \"' . $registry . '\"] }" > /etc/docker/daemon.json');
+    assert_script_run('cat /etc/docker/daemon.json');
+    systemctl('restart docker');
+    record_info "setup $self->runtime", "deamon.json ready";
+}
+
+package containers::engine::podman;
+use Mojo::Base 'containers::engine';
+use testapi;
+use containers::utils qw(registry_url);
+use utils qw(systemctl file_content_replace);
+has runtime => "podman";
+
+sub configure_insecure_registries {
+    my ($self) = shift;
+    my $registry = registry_url();
+
+    assert_script_run "curl " . data_url('containers/registries.conf') . " -o /etc/containers/registries.conf";
+    assert_script_run "chmod 644 /etc/containers/registries.conf";
+    file_content_replace("/etc/containers/registries.conf", REGISTRY => $registry);
+}
+
+
+package containers::engine::buildah;
+use Mojo::Base 'containers::engine';
+use testapi;
+use containers::utils qw(registry_url);
+use utils qw(systemctl file_content_replace);
+has runtime => "buildah";
+
+sub cleanup_system_host {
+    my ($self, $assert) = @_;
+    $assert // 1;
```

Comment:
I always thought the syntax is `//=`

Reply:
you can check the documentation. i use both

Reply:
> you can check the documentation. i use both
Nope, 

```
use strict;
use warnings;
use feature 'say';

sub function {
	my ($arg) = @_;
	$arg // 1;
	say $arg;
}

function();
function('hello');
```

>perl test.pl 
> Use of uninitialized value $arg in say at test.pl line 8.
> hello

in this case, what you want to do is what @jlausuch pointed out:

```
$arg //= 1;
```

Reply:
*this change is ready for review*